File: MasterNode/src/main/java/com/yahoo/sdvornik/master/Merger.java
package com.yahoo.sdvornik.master;

import com.yahoo.sdvornik.Constants;
import fj.Unit;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.channels.WritableByteChannel;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardOpenOption;
import java.util.EnumSet;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.LinkedBlockingDeque;

public final class Merger {

    private final static Logger log = LoggerFactory.getLogger(MasterTask.class.getName());


    private final Map<String, LinkedBlockingDeque<long[]>> dequeMap = new HashMap<>();
    private final ExecutorService executor = Executors.newFixedThreadPool(1);

    private final fj.data.List<String> idList;
    private final long numberOfKeys;
    private final int totalKeysInOneGeneration;
    private final Path pathToResult;

    private final fj.F0<Unit> before;
    private final fj.F<String, Unit> onError;
    private final fj.F0<Unit> onSuccess;

    public Merger(
            fj.data.List<String> idList,
            long numberOfKeys,
            int totalKeysInOneGeneration,
            Path pathToResult,
            fj.F0<Unit> before,
            fj.F<String, Unit> onError,
            fj.F0<Unit> onSuccess
    ) {
        this.idList = idList;
        this.numberOfKeys = numberOfKeys;
        this.totalKeysInOneGeneration = totalKeysInOneGeneration;
        this.pathToResult = pathToResult;
        this.before = before;
        this.onError = onError;
        this.onSuccess = onSuccess;
    }

    private Runnable mergeTask = new Runnable() {

        @Override
        public void run() {

            if(before!=null) before.f();

            int maxGeneration = ((int)numberOfKeys/totalKeysInOneGeneration +
                    (numberOfKeys%totalKeysInOneGeneration == 0 ? 0:1));

            try {
                multiMergeAndSave(idList.array(String[].class),  numberOfKeys, totalKeysInOneGeneration);
                if(onSuccess!=null) onSuccess.f();
            }
            catch(InterruptedException e) {
                if(onError!=null) onError.f("Unexpected InterruptedException");
            }
            dequeMap.clear();
        }
    };

    public void init() {
        idList.foreach(
                id -> {
                    dequeMap.put(id, new LinkedBlockingDeque<long[]>());
                    return Unit.unit();
                }
        );
        executor.execute(mergeTask);
        log.info("Init Merger");
    }

    public void shutdownNow() {
        dequeMap.clear();
        executor.shutdownNow();
    }

    public void multiMergeAndSave(String[] id, long numberOfKeys, int totalKeysInOneGeneration) throws InterruptedException {

        int maxGeneration = ((int)numberOfKeys/totalKeysInOneGeneration +
                (numberOfKeys%totalKeysInOneGeneration == 0 ? 0:1));

        long[][] multiArr = new long[id.length][];
        int[] curIndex = new int[id.length];
        int[] curGeneration = new int[id.length];

        int curNumberOfArrWithMinValue = -1;
        int generation = 0;

        for(int i = 0; i< id.length; ++i) {
            multiArr[i] = dequeMap.get(id[i]).takeFirst();
            curGeneration[i]=1;
        }

        try (WritableByteChannel writableByteChannel = Files.newByteChannel(pathToResult,
                EnumSet.of(StandardOpenOption.WRITE, StandardOpenOption.APPEND))) {

            ByteBuffer buffer = ByteBuffer.allocate(totalKeysInOneGeneration*Long.BYTES);

            while(generation < maxGeneration) {
                int mergeArrLength = (int)(generation < maxGeneration - 1 ?
                        totalKeysInOneGeneration : numberOfKeys - totalKeysInOneGeneration*generation);
                long[] mergeArr = new long[mergeArrLength];
                for (int i = 0; i < mergeArrLength; ++i) {
                    long curMinValue = Long.MAX_VALUE;

                    for (int k = 0; k < multiArr.length; ++k) {
                        if (curIndex[k] < multiArr[k].length && multiArr[k][curIndex[k]] < curMinValue) {
                            curMinValue = multiArr[k][curIndex[k]];
                            curNumberOfArrWithMinValue = k;
                        }
                    }
                    mergeArr[i] = curMinValue;
                    ++curIndex[curNumberOfArrWithMinValue];

                    if (curIndex[curNumberOfArrWithMinValue] == multiArr[curNumberOfArrWithMinValue].length) {

                        if(curGeneration[curNumberOfArrWithMinValue]<maxGeneration) {

                            LinkedBlockingDeque<long[]> deque = dequeMap.get(id[curNumberOfArrWithMinValue]);

                            multiArr[curNumberOfArrWithMinValue] = deque.takeFirst();

                            curIndex[curNumberOfArrWithMinValue] = 0;
                            ++curGeneration[curNumberOfArrWithMinValue];
                        }
                    }
                }

                for (int i = 0; i < mergeArr.length; ++i) {
                    buffer.putLong(mergeArr[i]);
                }
                buffer.flip();
                writableByteChannel.write(buffer);
                buffer.clear();

                ++generation;
            }
        }
        catch(IOException e) {
            log.error("Unexpected error while writing to file.", e);
        }
    }

    public void putArrayInQueue(String id, int numberOfChunk, long[] sortedArr) {
        LinkedBlockingDeque<long[]> deque = dequeMap.get(id);
        deque.addLast(sortedArr);
    }
}

File: master/src/main/kotlin/com/yahoo/sdvornik/master/Merger.kt
package com.yahoo.sdvornik.master

import fj.Unit
import org.slf4j.LoggerFactory
import java.io.IOException
import java.nio.ByteBuffer
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardOpenOption
import java.util.*
import java.util.concurrent.Executors
import java.util.concurrent.LinkedBlockingDeque

/**
 * @author Serg Dvornik <sdvornik@yahoo.com>
 */
class Merger(
        private val idList: fj.data.List<String>,
        private val numberOfKeys: Long,
        private val totalKeysInOneGeneration: Int,
        private val pathToResult: Path,
        private val before: fj.F0<Unit>?,
        private val onError: fj.F<String, Unit>?,
        private val onSuccess: fj.F0<Unit>?
) {


    private val dequeMap = HashMap<String, LinkedBlockingDeque<LongArray>>()
    private val executor = Executors.newFixedThreadPool(1)

    private val mergeTask = Runnable {
        before?.f()

        val maxGeneration = numberOfKeys.toInt() / totalKeysInOneGeneration + if (numberOfKeys.toInt() % totalKeysInOneGeneration == 0) 0 else 1

        try {
            multiMergeAndSave(idList.array(Array<String>::class.java), numberOfKeys, totalKeysInOneGeneration)
            onSuccess?.f()
        } catch (e: InterruptedException) {
            onError?.f("Unexpected InterruptedException")
        }

        dequeMap.clear()
    }

    fun init() {
        idList.foreach { id ->
            dequeMap.put(id, LinkedBlockingDeque())
            Unit.unit()
        }
        executor.execute(mergeTask)
        log.info("Init Merger")
    }

    fun shutdownNow() {
        dequeMap.clear()
        executor.shutdownNow()
    }

    @Throws(InterruptedException::class)
    fun multiMergeAndSave(id: Array<String>, numberOfKeys: Long, totalKeysInOneGeneration: Int) {

        val maxGeneration = numberOfKeys.toInt() / totalKeysInOneGeneration + if (numberOfKeys.toInt() % totalKeysInOneGeneration == 0) 0 else 1

        val multiArr = arrayOfNulls<LongArray>(id.size)
        val curIndex = IntArray(id.size)
        val curGeneration = IntArray(id.size)

        var curNumberOfArrWithMinValue = -1
        var generation = 0

        for (i in id.indices) {
            multiArr[i] = dequeMap[id[i]]?.takeFirst()
            curGeneration[i] = 1
        }

        try {
            Files.newByteChannel(pathToResult,
                    EnumSet.of(StandardOpenOption.WRITE, StandardOpenOption.APPEND)).use { writableByteChannel ->

                val buffer = ByteBuffer.allocate(totalKeysInOneGeneration * java.lang.Long.BYTES)

                while (generation < maxGeneration) {
                    val mergeArrLength: Int = if (generation < maxGeneration - 1) totalKeysInOneGeneration
                    else numberOfKeys.toInt() - totalKeysInOneGeneration * generation
                    val mergeArr = LongArray(mergeArrLength)
                    for (i in 0..mergeArrLength - 1) {
                        var curMinValue = java.lang.Long.MAX_VALUE

                        for (k in multiArr.indices) {
                            val multiArrK: LongArray? = multiArr[k]
                            if (curIndex[k] < multiArrK!!.size && multiArrK[curIndex[k]] < curMinValue) {
                                curMinValue = multiArrK[curIndex[k]]
                                curNumberOfArrWithMinValue = k
                            }
                        }
                        mergeArr[i] = curMinValue
                        ++curIndex[curNumberOfArrWithMinValue]

                        if (curIndex[curNumberOfArrWithMinValue] == multiArr[curNumberOfArrWithMinValue]?.size) {

                            if (curGeneration[curNumberOfArrWithMinValue] < maxGeneration) {

                                val deque = dequeMap[id[curNumberOfArrWithMinValue]]

                                multiArr[curNumberOfArrWithMinValue] = deque?.takeFirst()

                                curIndex[curNumberOfArrWithMinValue] = 0
                                ++curGeneration[curNumberOfArrWithMinValue]
                            }
                        }
                    }

                    for (i in mergeArr.indices) {
                        buffer.putLong(mergeArr[i])
                    }
                    buffer.flip()
                    writableByteChannel.write(buffer)
                    buffer.clear()

                    ++generation
                }
            }
        } catch (e: IOException) {
            log.error("Unexpected error while writing to file.", e)
        }

    }

    fun putArrayInQueue(id: String, numberOfChunk: Int, sortedArr: LongArray) {
        val deque = dequeMap[id]
        deque?.addLast(sortedArr)
    }

    companion object {

        private val log = LoggerFactory.getLogger(MasterTask::class.java)
    }
}
